elastic/beats
# Add leader election for autodiscover

## What does this PR do?
Implements leader election as part of kubernetes autodiscover provider. With this, when the beat is elected as leader it will start cluster scope metricsets.

Under the hood, all leader candidates will try to gain the lock on Lease object (https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.18/#lease-v1-coordination-k8s-io). When the lock is gained then the configured cluster wide mericsets will get started.

## Why is it important?
To get rid of `Deployment` procedure for the singleton Metricbeat instance which is needed for the cluster scope metricsets.
Sample configuration that will start `state_node` metricset when a leadership is gained:

```
metricbeat.autodiscover:
  providers:
    - type: kubernetes
      scope: cluster
      node: ${NODE_NAME}
      unique: true
      identifier: leaderelectionmetricbeat
      templates:
        - config:
            - module: kubernetes
              hosts: ["kube-state-metrics:8080"]
              period: 10s
              add_metadata: true
              metricsets:
                - state_node
```

Then we can monitor the `lease` to check which provider holds the lock:
```console
kubectl describe lease beats-cluster-leader
Name:         beats-cluster-leader
Namespace:    default
Labels:       <none>
Annotations:  <none>
API Version:  coordination.k8s.io/v1
Kind:         Lease
Metadata:
  Creation Timestamp:  2020-07-30T14:48:16Z
  Resource Version:    357033
  Self Link:           /apis/coordination.k8s.io/v1/namespaces/default/leases/beats-cluster-leader
  UID:                 ef255352-3d30-42f9-b27a-3a3e5dcb80dd
Spec:
  Acquire Time:            2020-07-30T14:51:00.592028Z
  Holder Identity:         beats-leader-leaderelectionmetricbeat
  Lease Duration Seconds:  15
  Lease Transitions:       2
  Renew Time:              2020-07-30T14:51:00.597852Z
Events:                    <none>
```

## Checklist



## How to test this PR locally
1. Prepare a multinode k8s cluster (for instance on gke)
2. Deploy Metricbeat as a Daemonset on the cluster with leader election config:
```
metricbeat.autodiscover:
  providers:
    - type: kubernetes
      scope: cluster
      node: ${NODE_NAME}
      unique: true
      leader_lease: leaderelectionmetricbeat
      templates:
        - config:
            - module: kubernetes
              hosts: ["kube-state-metrics:8080"]
              period: 10s
              add_metadata: true
              metricsets:
                - state_node
```
3. Make sure that events for `state_node` are being collected from only one metricbeat instance. You can the in the logs and verify that only one metricbeat has gained the lock so far.
4. Monitor the `lease` object to keep track of the lock holder: `watch kubectl describe lease beats-cluster-leader`
5. Stop the Metricbeat instance that currently holds the lock and make sure that another one takes over and we keep getting events from `state-node`. Make sure that the `lease` changed holder (logs will also let us know that a new metricset is getting started)
6. Also test without setting `identifier` and check that it takes the default value properly like `metricbeat-cluster-leader`.

7. Check with old autodiscover (template based/hints based) for possible regressions.

## Related issues

- Relates https://github.com/elastic/beats/issues/19731
